extensions/amp-pan-zoom/1.0/component-ts.tsx

```diff
@@ -0,0 +1,268 @@
+
+  useImperativeHandle(
+    ref,
+    () => /** @type {!BentoPanZoom.PanZoomApi} */ ({
+      transform: (scale: number, x: number, y: number) => {
+        actions.transform({
+          scale,
+          posX: x,
+          posY: y,
+        });
+      },
+    }),
+    [actions]
+  );
+
+  const panZoomStyles = {
+    transformOrigin:
+      px(-state.contentOffset.x) + ' ' + px(-state.contentOffset.y),
+    transform: translate(state.posX, state.posY) + cssScale(state.scale),
+  };
+
+  const initialStateRef = useRef(state);
+  const lastTapTime = useRef(-DOUBLE_TAP_TIME);
+  const handlers: Partial<UserHandlers> = {
+    onDragStart() {
+      if (isPannable) {
+        actions.draggingStart();
+        initialStateRef.current = state;
+      }
+    },
+    onDragEnd() {
+      actions.draggingRelease();
+    },
+    onDrag(ev) {
+      // Taps come through the `onDrag` handler:
+      if (ev.tap) {
+        const isDoubleTap =
+          ev.startTime - lastTapTime.current <= DOUBLE_TAP_TIME;
+        if (isDoubleTap) {
+          // Double tap:
+          const [clientX, clientY] = ev.xy;
+          const anchor = getElementPosition(
+            clientX,
+            clientY,
+            containerRef.current!
+          );
+          actions.updateScale(anchor);
+          lastTapTime.current = -DOUBLE_TAP_TIME; // Prevent triple-taps
+        } else {
+          // Single tap
+          lastTapTime.current = ev.startTime;
+        }
+        return;
+      }
+
+      // Let's pan!
+      if (!isPannable) {
+        return;
+      }
+      const [deltaX, deltaY] = ev.movement;
+
+      const {posX: initialX, posY: initialY} = initialStateRef.current;
+
+      actions.transform({
+        posX: initialX + deltaX,
+        posY: initialY + deltaY,
+      });
+    },
+
+    onPinchStart: () => {
+      actions.draggingStart();
+      initialStateRef.current = state;
+    },
+    onPinchEnd: () => {
+      actions.draggingRelease();
+    },
+    onPinch: (ev) => {
+      const [clientX, clientY] = ev.origin;
+      const [scale] = ev.movement;
+      const [initialClientX, initialClientY] = ev.memo || [clientX, clientY];
+
+      const {anchorX, anchorY} = getElementPosition(
+        initialClientX,
+        initialClientY,
+        containerRef.current!
+      );
+
+      const {
+        posX: initialX,
+        posY: initialY,
+        scale: initialScale,
+      } = initialStateRef.current;
+
+      const newScale = scale * initialScale;
+
+      // Reset position before scaling:
+      actions.transform({
+        posX: initialX,
+        posY: initialY,
+        scale: initialScale,
+      });
+      // Zoom in on anchor:
+      actions.updateScale({
+        anchorX,
+        anchorY,
+        scale: newScale,
+      });
+
+      if (ev.first) {
+        // Store the initial origin as `ev.memo`:
+        return ev.origin;
+      }
+    },
+  };
+  const bind = useGesture(handlers as GestureHandlers, gestureConfig);
```

Comment:
Is there a reason why handlers isn't of the type `GestureHandlers`?

Reply:
Yeah, it's strangely necessary, the types for this library are really problematic ... I think maybe they're conflicting with our particular TS config.  
Basically, `type GestureHandlers = Partial<NativeHandlers & UserHandlers>`, but these 2 types seem to have conflicts with each other.  So, I defined `const handlers: Partial<UserHandlers> = {...}` and that gives me all the strong types without conflicts.  But it requires this cast.